File: joptimize/test/src/examples/narrow/IntersectionReturn.scala
package joptimize.examples.narrow

object IntersectionReturn {
  def main(x: Int, y: Int): Int = {
    if (call(x, y) == Bar) 1 else 2
  }

  def call(x: Int, y: Int): Foo = {
    if (x > y) Bar
    else Qux
  }

  trait Foo{
    def inc(n: Int): Int
  }
  trait Foo1 extends Foo
  trait Foo2 extends Foo
  object Bar extends Foo1 with Foo2{
    def inc(n: Int) = n + 1
  }
  object Qux extends Foo1 with Foo2{
    def inc(n: Int) = n + 2
  }

}

File: joptimize/test/src/examples/narrow/IntersectionReturn.java
package joptimize.examples.narrow;

class IntersectionReturn {
    static int main(int x, int y) {
        return call(x, y) instanceof Bar ? 1 : 2;
    }

    static Foo call(int x, int y) {
        return x > y ? new Bar() : new Qux();
    }

    interface Foo {
        int inc(int n);
    }

    interface Foo1 extends Foo {
    }

    interface Foo2 extends Foo {
    }

    static class Bar implements Foo1, Foo2 {
        public int inc(int n) {
            return n + 1;
        }
    }

    static class Qux implements Foo1, Foo2 {
        public int inc(int n) {
            return n + 2;
        }
    }
}

